### PR TITLE
ADEN-10310 Context targeting refactor 

### PR DESCRIPTION
**Pull request by lnawrotek**

### CODE REVIEW

File: src/ad-bidders/prebid/adapters/appnexus.ts [hunk 1]
```diff
@@ -74,7 +74,7 @@ export class Appnexus extends PrebidAdapter {
 		let placement = position;
 
 		if (position === 'mobile') {
-			const vertical = context.get('targeting.mappedVerticalName');
+			const vertical = targetingService.getAll<TargetingData>().mappedVerticalName;
```
Comment (dmnsobczak) [inline src/ad-bidders/prebid/adapters/appnexus.ts:77]: We can get rid of `getAll()` and use only `get(null)` to get whole targeting

File: src/platforms/news-and-ratings/shared/setup/context/targeting/news-and-ratings-targeting.setup.ts [hunk 2]
```diff
@@ -19,8 +20,8 @@ export class NewsAndRatingsTargetingSetup implements DiProcess {
 
 		this.setSlotLevelTargeting(targeting, customConfig);
 
-		context.set('targeting', {
-			...context.get('targeting'),
+		targetingService.extend({
+			...targetingService.getAll(),
```
Comment (dmnsobczak) [inline src/platforms/news-and-ratings/shared/setup/context/targeting/news-and-ratings-targeting.setup.ts:24]: Missing a type

File: src/platforms/shared/context/targeting/ucp-targeting.setup.ts [hunk 3]
```diff
@@ -51,11 +53,11 @@ export class UcpTargetingSetup implements DiProcess {
 
 		if (context.get('wiki.targeting.wikiIsTop1000')) {
 			context.set('custom.wikiIdentifier', '_top1k_wiki');
-			context.set('custom.dbNameForAdUnit', context.get('targeting.s1'));
+			context.set('custom.dbNameForAdUnit', targetingService.getAll<TargetingData>().s1);
```
Comment (dmnsobczak) [inline src/platforms/shared/context/targeting/ucp-targeting.setup.ts:56]: Think about removing generics

File: src/platforms/shared/tracking/ad-size-tracker.ts [hunk 4]
```diff
@@ -32,7 +32,7 @@ export class AdSizeTracker {
 				slot_id: slot.getUid() || '',
 				creative_id: slot.creativeId || '',
 				line_item_id: slot.lineItemId || '',
-				rv: slot.getConfigProperty('targeting.rv') || '',
+				rv: slot.getTargetingConfigProperty('rv') || '',
```
Comment (dmnsobczak) [inline src/platforms/shared/tracking/ad-size-tracker.ts:35]: `slot.getTargetingProperty('rv')`

File: src/core/providers/gpt-targeting.ts [hunk 5]
```diff
@@ -17,7 +17,7 @@ export function setupGptTargeting(): void {
 	}
 
 	function setTargetingFromContext(): void {
```
Comment (dmnsobczak) [inline src/core/providers/gpt-targeting.ts:19]: `Context` can be removed

File: src/core/services/slot-repeater.ts [hunk 6]
```diff
@@ -42,7 +43,11 @@ function repeatSlot(adSlot: AdSlot): boolean {
 					? buildString(repeatConfig.updateProperties[key], newSlotDefinition)
 					: repeatConfig.updateProperties[key];
 
-			context.set(`slots.${slotName}.${key}`, value);
+			if (key.startsWith('targeting.')) {
```
Comment (dmnsobczak) [inline src/core/services/slot-repeater.ts:46]: Keys in adUnits templates can be changed to get rid of it

File: src/core/services/slot-targeting-service.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+	constructor() {
+		window.ads.slotTargeting = debug.isDebugMode() ? this.slotTargeting : {};
+	}
+
+	extend(slotName: string, newTargeting: SlotTargeting): void {
+		this.slotTargeting[slotName] = this.slotTargeting[slotName] || ({} as SlotTargeting);
+
+		this.slotTargeting[slotName] = Object.assign(this.slotTargeting[slotName], newTargeting);
+		window.ads.slotTargeting = debug.isDebugMode() ? this.slotTargeting : {};
```
Comment (dmnsobczak) [inline src/core/services/slot-targeting-service.ts:44]: JS probably copies the object, not passing the reference - we can find out why

Comment (dmnsobczak) [reply src/core/services/slot-targeting-service.ts:44]: `window.ads.slotTargeting` can be changed to callback

File: src/core/services/targeting-service.ts [hunk 8]
```diff
@@ -0,0 +1,116 @@
+import { debug } from './debug';
+
+type ChangeCallback = (key: string | null, value: any) => void;
+
+interface TargetingObject {
+	[key: string]: any;
+}
+
+export interface TargetingData {
```
Comment (dmnsobczak) [inline src/core/services/targeting-service.ts:9]:
```TargetingData = Audigent & Something & {
   ...
}```

File: src/platforms/ucp-desktop/ucp-desktop-platform.ts [hunk 9]
```diff
@@ -35,6 +36,7 @@ export class UcpDesktopPlatform {
 		// Config
 		this.pipeline.add(
 			() => context.extend(basicContext),
+			() => targetingService.extend(targetingData),
```
Comment (dmnsobczak) [inline src/platforms/ucp-desktop/ucp-desktop-platform.ts:39]: Discussed, to check

File: src/platforms/ucp-desktop/ad-context.ts [hunk 10]
```diff
@@ -123,3 +123,9 @@ export const basicContext = {
 		},
 	},
 };
+
+export const targetingData = {
```
Comment (dmnsobczak) [inline src/platforms/ucp-desktop/ad-context.ts:127]: Discussed, to check

File: @types/ads/index.d.ts [hunk 11]
```diff
@@ -1,6 +1,8 @@
 interface MediaWikiAds {
 	adEngineVersion: string;
 	adContext: any;
+	adTargeting: any;
+	slotTargeting: any;
```
Comment (nandy-andy) [inline @types/ads/index.d.ts:5]: What do you think of `pageTargeting` and `slotTargeting`? I wonder how `adTargeting` is different from `pageTargeting` we use in GPT: https://developers.google.com/publisher-tag/guides/key-value-targeting

File: spec/ad-bidders/prebid/adapters/appnexus.spec.ts [hunk 12]
```diff
@@ -1,8 +1,20 @@
 import { Appnexus } from '@wikia/ad-bidders/prebid/adapters/appnexus';
-import { context } from '@wikia/core';
+import { TargetingService, targetingService } from '@wikia/core';
```
Comment (nandy-andy) [inline spec/ad-bidders/prebid/adapters/appnexus.spec.ts:2]: This looks strange. Consider importing just `TargetingService` and creating its instance in the test 🤔 I think once we merge the changes we'd be able to use the dependency-injection Damian worked on during karpik.

Comment (nandy-andy) [reply spec/ad-bidders/prebid/adapters/appnexus.spec.ts:2]: Oh, but maybe it won't work as we need a singleton we test in the cases below 🤔

File: src/core/utils/string-builder.ts [hunk 13]
```diff
@@ -25,6 +25,10 @@ class StringBuilder {
 					}
 				}
 
+				if (typeof value === 'undefined' && key.startsWith('targeting.')) {
```
Comment (nandy-andy) [inline src/core/utils/string-builder.ts:28]: This is the 2nd time of `startsWith('targeting.')` usage. Consider extracting it as a part of the `TargetingService`.

File: src/platforms/f2/f2-platform.ts [hunk 14]
```diff
@@ -38,11 +38,11 @@ export class F2Platform {
 		// Config
 		this.pipeline.add(
 			() => context.extend(basicContext),
+			() => context.set('state.isMobile', f2env.isPageMobile),
```
Comment (nandy-andy) [inline src/platforms/f2/f2-platform.ts:41]: What is the reason for this change? 🤔 I think it makes sense to have it here (as soon as possible) but I wonder if it was the only reason 😊

File: src/platforms/f2/setup/context/targeting/f2-targeting.setup.ts [hunk 15]
```diff
@@ -22,26 +23,28 @@ export class F2TargetingSetup implements DiProcess {
...
+			uap_c: 'none',
 			s0: 'fandom',
 			s1: '_fandom',
 			s2: this.f2State.pageType === 'topic' ? 'vertical' : this.f2State.pageType,
 			rating: 'esrb:teen',
 			labrador: this.cacheStorage.mapSamplingResults(
 				this.instantConfig.get('icLABradorGamKeyValues'),
 			),
-			is_mobile: utils.client.isMobileSkin(this.f2Env.skinName) ? '1' : '0',
+			is_mobile: context.get('state.isMobile') ? '1' : '0',
```
Comment (nandy-andy) [inline src/platforms/f2/setup/context/targeting/f2-targeting.setup.ts:47]: We could also use here `utils.client.isMobile()` 🤔 But I think basing on the response from backend is more reliable as someone can open a desktop view/skin on a mobile device and the other way around 👍

File: src/platforms/shared/tracking/data-warehouse.ts [hunk 16]
```diff
@@ -47,10 +47,10 @@ export class DataWarehouseTracker {
 			c: context.get('wiki.wgCityId') || 'unknown',
 			ck: context.get('wiki.dsSiteKey') || 'unknown',
 			lc: context.get('wiki.wgUserLanguage') || 'unknown',
-			s: context.get('targeting.skin') || 'unknown',
+			s: targetingService.get('skin') || 'unknown',
 			ua: window.navigator.userAgent,
 			u: trackingOptIn.isOptedIn() ? context.get('userId') || 0 : -1,
-			a: context.get('targeting.artid') || -1,
+			a: targetingService.get('artid') ? parseInt(targetingService.get('artid')) : -1,
```
Comment (dmnsobczak) [inline src/platforms/shared/tracking/data-warehouse.ts:53]:
```suggestion
			a: parseInt(targetingService.get('artid') || -1),
```

**Review by dmnsobczak — APPROVED** [review]

File: src/platforms/f2/f2-platform.ts (hunk 14 above)
Comment (dmnsobczak) [reply src/platforms/f2/f2-platform.ts:41]: + Code cleanup :D

File: @types/ads/index.d.ts (hunk 11 above)
Comment (dmnsobczak) [reply @types/ads/index.d.ts:5]: Also: `slotTargeting` is not being set on `window.ads`

File: src/platforms/f2/setup/context/targeting/f2-targeting.setup.ts (hunk 15 above)
Comment (dmnsobczak) [reply src/platforms/f2/setup/context/targeting/f2-targeting.setup.ts:47]: Yep. Also if we already have this value in the context - we can use it :)

File: @types/ads/index.d.ts (hunk 11 above)
Comment (lnawrotek) [reply @types/ads/index.d.ts]: Done
- slotTargeting - removed
- adTargeting > pageTargeting

File: spec/ad-bidders/prebid/adapters/appnexus.spec.ts (hunk 12 above)
Comment (lnawrotek) [reply spec/ad-bidders/prebid/adapters/appnexus.spec.ts:2]: I need this because I create the stub 
`let targetingServiceStub: SinonStubbedInstance<TargetingService>;

beforeEach(() => {
  sandbox = createSandbox();
  targetingServiceStub = sandbox.stub(targetingService);
});
`

File: src/core/utils/string-builder.ts (hunk 13 above)
Comment (lnawrotek) [reply src/core/utils/string-builder.ts:28]: In my opinion, we don't need to extract it

**Review by mfarul — DISMISSED** [review]

**Review by dmnsobczak — APPROVED** [review]

**Review by dmnsobczak — APPROVED** [review]

**Review by dmnsobczak — APPROVED** [review]

**Comment by dmnsobczak** [comment]
@kwfandom Be advised: merging dev usually requires some manual work to cover new areas of code with a `targetingService` if somebody will try to use old approach `context.set('targeting', xxx)`.
cc @lnawrotek

**Comment by kwfandom** [comment]
@dmnsobczak so we definitely have to talk on Monday

**Review by dmnsobczak — APPROVED** [review]

**Review by dmnsobczak — APPROVED** [review]
Approving in the name of @lnawrotek :D
